File: Task4/Task4.cpp
#include <iostream>
#include <vector>
#include <string>
#include <map>
#include <set>
#include <sstream>
#include <algorithm>

using namespace std;

// Поиск позиции кандидата в предпочтении избирателя (0 = лучший)
int findPosition(const vector<string>& pref, const string& cand) {
    for (size_t i = 0; i < pref.size(); ++i)
        if (pref[i] == cand) return i;
    return -1;
}

// Метод Борда
map<string, int> bordaScore(const vector<string>& candidates, const vector<vector<string>>& votes) {
    int n = candidates.size();
    map<string, int> scores;
    for (const string& c : candidates) scores[c] = 0;

    for (const auto& vote : votes) {
        for (size_t pos = 0; pos < vote.size(); ++pos) {
            const string& cand = vote[pos];
            scores[cand] += (n - 1 - pos);
        }
    }
    return scores;
}

// Метод Кондорсе
string condorcetWinner(const vector<string>& candidates, const vector<vector<string>>& votes) {
    int n = candidates.size();
    map<string, map<string, int>> beat;
    for (const string& a : candidates)
        for (const string& b : candidates)
            if (a != b) beat[a][b] = 0;

    for (const auto& vote : votes) {
        for (size_t i = 0; i < vote.size(); ++i) {
            for (size_t j = i + 1; j < vote.size(); ++j) {
                beat[vote[i]][vote[j]]++;
            }
        }
    }

    for (const string& a : candidates) {
        bool winsAll = true;
        for (const string& b : candidates) {
            if (a == b) continue;
            if (beat[a][b] <= beat[b][a]) {
                winsAll = false;
                break;
            }
        }
        if (winsAll) return a;
    }
    return "";
}

void printScores(const map<string, int>& scores) {
    cout << "\n--- Результаты по методу Борда ---\n";
    for (const auto& p : scores)
        cout << p.first << ": " << p.second << " очков\n";
    int maxScore = -1;
    string winner;
    for (const auto& p : scores) {
        if (p.second > maxScore) {
            maxScore = p.second;
            winner = p.first;
        }
    }
    cout << "\nПобедитель по Борду: " << winner << " (" << maxScore << " очков)\n";
}

int main() {
    int nCandidates;
    cout << "Введите количество кандидатов: ";
    cin >> nCandidates;
    cin.ignore();
    vector<string> candidates(nCandidates);
    cout << "Введите имена кандидатов (каждое с новой строки):\n";
    for (int i = 0; i < nCandidates; ++i) {
        getline(cin, candidates[i]);
    }

    int nVoters;
    cout << "Введите количество избирателей: ";
    cin >> nVoters;
    cin.ignore();

    vector<vector<string>> votes(nVoters);
    cout << "Для каждого избирателя введите порядок предпочтения (имена кандидатов в порядке убывания предпочтения, через пробел):\n";
    for (int i = 0; i < nVoters; ++i) {
        cout << "Избиратель " << i+1 << ": ";
        string line;
        getline(cin, line);
        stringstream ss(line);
        string name;
        while (ss >> name) {
            votes[i].push_back(name);
        }
        // Если ввели меньше, чем кандидатов, то запрашиваем остальных
        while (votes[i].size() < candidates.size()) {
            string extra;
            cin >> extra;
            votes[i].push_back(extra);
        }
        // Очистка буфера после возможного дополнительного ввода
        if (votes[i].size() > candidates.size()) {
            cin.ignore();
        }
    }

    auto borda = bordaScore(candidates, votes);
    printScores(borda);

    string cWinner = condorcetWinner(candidates, votes);
    cout << "\n--- Результаты по методу Кондорсе ---\n";
    if (cWinner.empty())
        cout << "Победитель по Кондорсе не определён (цикл Кондорсе).\n";
    else
        cout << "Победитель по Кондорсе: " << cWinner << "\n";

    return 0;
}

File: Task4/Task4.rs
use std::collections::BTreeMap;
use std::io::{self, Write};

// Поиск позиции кандидата в предпочтении избирателя (0 = лучший)
#[allow(dead_code)]
fn find_position(pref: &[String], cand: &str) -> Option<usize> {
    pref.iter().position(|c| c == cand)
}

// Метод Борда
fn borda_score(candidates: &[String], votes: &[Vec<String>]) -> BTreeMap<String, i32> {
    let n = candidates.len() as i32;
    let mut scores = BTreeMap::new();
    for c in candidates {
        scores.insert(c.clone(), 0);
    }
    for vote in votes {
        for (pos, cand) in vote.iter().enumerate() {
            let points = n - 1 - pos as i32;
            *scores.get_mut(cand).unwrap() += points;
        }
    }
    scores
}

// Метод Кондорсе
fn condorcet_winner(candidates: &[String], votes: &[Vec<String>]) -> Option<String> {
    // Инициализируем матрицу парных побед
    let mut beat = BTreeMap::new();
    for a in candidates {
        let mut row = BTreeMap::new();
        for b in candidates {
            if a != b {
                row.insert(b.clone(), 0);
            }
        }
        beat.insert(a.clone(), row);
    }

    // Подсчёт предпочтений
    for vote in votes {
        for i in 0..vote.len() {
            for j in i + 1..vote.len() {
                let winner = &vote[i];
                let loser = &vote[j];
                *beat.get_mut(winner).unwrap().get_mut(loser).unwrap() += 1;
            }
        }
    }

    // Проверка кандидата, побеждающего всех
    for a in candidates {
        let mut wins_all = true;
        for b in candidates {
            if a == b {
                continue;
            }
            let a_beats_b = beat[a][b];
            let b_beats_a = beat[b][a];
            if a_beats_b <= b_beats_a {
                wins_all = false;
                break;
            }
        }
        if wins_all {
            return Some(a.clone());
        }
    }
    None
}

fn print_scores(scores: &BTreeMap<String, i32>) {
    println!("\n--- Результаты по методу Борда ---");
    for (name, score) in scores {
        println!("{}: {} очков", name, score);
    }
    let winner = scores.iter().max_by_key(|&(_, &score)| score).unwrap();
    println!("\nПобедитель по Борду: {} ({} очков)", winner.0, winner.1);
}

fn main() {
    // Ввод кандидатов
    print!("Введите количество кандидатов: ");
    io::stdout().flush().unwrap();
    let mut input = String::new();
    io::stdin().read_line(&mut input).unwrap();
    let n_candidates: usize = input.trim().parse().expect("Некорректное число");
    let mut candidates = Vec::with_capacity(n_candidates);
    println!("Введите имена кандидатов (каждое с новой строки):");
    for _ in 0..n_candidates {
        let mut name = String::new();
        io::stdin().read_line(&mut name).unwrap();
        candidates.push(name.trim().to_string());
    }

    // Ввод избирателей
    print!("Введите количество избирателей: ");
    io::stdout().flush().unwrap();
    input.clear();
    io::stdin().read_line(&mut input).unwrap();
    let n_voters: usize = input.trim().parse().expect("Некорректное число");
    let mut votes = Vec::with_capacity(n_voters);
    println!("Для каждого избирателя введите порядок предпочтения (имена кандидатов в порядке убывания предпочтения, через пробел):");
    for i in 0..n_voters {
        print!("Избиратель {}: ", i + 1);
        io::stdout().flush().unwrap();
        let mut line = String::new();
        io::stdin().read_line(&mut line).unwrap();
        let preferences: Vec<String> = line
            .split_whitespace()
            .map(|s| s.to_string())
            .collect();
        // Если ввели меньше, чем кандидатов, запрашиваем остальных
        let mut pref = preferences;
        while pref.len() < candidates.len() {
            let mut extra = String::new();
            io::stdin().read_line(&mut extra).unwrap();
            pref.extend(extra.split_whitespace().map(|s| s.to_string()));
        }
        votes.push(pref);
    }

    // Расчёт и вывод
    let borda = borda_score(&candidates, &votes);
    print_scores(&borda);

    let c_winner = condorcet_winner(&candidates, &votes);
    println!("\n--- Результаты по методу Кондорсе ---");
    match c_winner {
        Some(winner) => println!("Победитель по Кондорсе: {}", winner),
        None => println!("Победитель по Кондорсе не определён (цикл Кондорсе)."),
    }
}